Use non-atomic flushing with block replay

sipa opened the pull request:
This implements an alternative solution to the flush-time memory usage peak, suggested by @gmaxwell.

Instead of relying on using atomic batch writes in LevelDB for the chainstate, we rely on the fact that we have an external log of updates to it already (called the blockchain).

This patch adds an extra "head blocks" to the chainstate, which gives the range of blocks for writes may be incomplete. At the start of a flush, we write this record, write the dirty dbcache entries in 16 MiB batches, and at the end we remove the heads record again. If it is present at startup it means we crashed during flush, and we rollback/roll forward blocks inside of it to get a consistent tip on disk before proceeding.

If a flush completes succesfully, the resulting database is compatible with previous versions (down to 0.8). If the node crashes in the middle of a flush, a version of the code with this patch is needed to recovery.

sipa commented:
This badly needs testing, but I'm not sure how to simulate crashes in the middle of flushing.

laanwj commented:
Cool!

> but I'm not sure how to simulate crashes in the middle of flushing 

I'll get to that :)

sipa committed:
Random db flush crash simulator

sipa commented:
Rebased, fixed a bug, and added a commit to allows simulating crashes after partial flushes.

laanwj commented on src/txdb.cpp:
```diff
@@ -74,6 +76,15 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
+                static FastRandomContext rng;
+                if (rng.rand32() % crash_simulate == 0) {
+                    LogPrintf("Simulating a crash. Goodbye.");
+                    sync();
```
This sync should be optional at least

sipa replied:
I've just removed the sync. It seems not needed for testing (I've done reindexes with hundreds of crashes in between).

gmaxwell commented:
contrib/devtools/check-doc.py is unhappy that you added new arguments without asking for permission from the argument gods.

gmaxwell commented on src/txdb.cpp:
```diff
@@ -44,10 +45,21 @@ uint256 CCoinsViewDB::GetBestBlock() const {
     CDBBatch batch(db);
     size_t count = 0;
     size_t changed = 0;
+    size_t batch_size = (size_t)GetArg("-dbbatchsize", nDefaultDbBatchSize) << 20;
```
If we have memory usage that is some multiple of this, perhaps the argument should be in the form of the actual usage rather than the batch size?

sipa replied:
Well the relevant constraint is the memory usage peak from allocating the batch, which depends on the batch memory usage, not dbcache memory usage. Also, I don't think anyone will need to change this property (except for tests, where it's very useful to get much more frequent partial flushes).

TheBlueMatt commented on src/init.cpp:
```diff
@@ -1476,6 +1479,12 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
+                    strLoadError = _("Unable to replay blocks. You will need to rebuild the databse using -reindex.");
+                    break;
+                }
+                pcoinsTip->SetBestBlock(pcoinsdbview->GetBestBlock()); // TODO: only initialize pcoinsTip after ReplayBlocks
```
Hmm, yes. Were you intending to do this after this PR? Can you just delete it and re-create it here? I feel like it may make sense to move the chainActive.Tip-setting from LoadBlockIndexDB to after this point.

Speaking of which, did you mean to add a PruneBlockIndexCandidates() to ReplayBlocks ala LoadBlockIndexDB?

sipa replied:
> Were you intending to do this after this PR?

Yes, I tried doing it inside the PR, but doing it properly requires a bit more shuffling around and refactoring, which I'd prefer to keep for later.

> Speaking of which, did you mean to add a PruneBlockIndexCandidates() to ReplayBlocks ala LoadBlockIndexDB?

ReplayBlocks doesn't touch the block index, so I don't think that would have any effect.

TheBlueMatt replied:
The PruneBlockIndexCandidates in LoadBlockIndexDB uses chainActive.Tip(), so I assumed it may need to be re-run with the new tip (though likely not a bug without it, just a should-do). I'm ok with cleaning this stuff up in a followup PR, but it seems less than ideal as-is right now.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3681,6 +3694,51 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    auto pindexUpto = mapBlockIndex[hashUpto];
+
+    int nHeight = 1; // Skip the genesis block
+    if (mapBlockIndex.count(hashBest)) {
```
Should we also fail here if hashBest has been written (ie is non-IsNull) but isnt in mapBlockIndex?

sipa replied:
I think that would be caught by other code we already have, but I've added it here.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1296,6 +1296,19 @@ void static InvalidBlockFound(CBlockIndex *pindex, const CValidationState &state
+                inputs.ModifyCoins(txin.prevout.hash)->Spend(txin.prevout.n);
+            }
+        }
+        inputs.ModifyNewCoins(tx->GetHash(), tx->IsCoinBase())->FromTx(*tx, nHeight);
```
Based on the comment above ModifyNewCoins I do not believe this works, we may need something new to capture the "maybe not fresh, but definitely fully overwrite in any case" case.

sipa replied:
Nice catch, fixed.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3681,6 +3694,51 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+        ReplayBlock(block, cache, pindex->nHeight);
+    }
+    cache.SetBestBlock(hashUpto);
+    chainActive.SetTip(pindexUpto);
```
It seems super weird to be acting entirely on non-globals, and then suddenly set a global here.

sipa replied:
Good point, fixed.

sipa commented on src/dbwrapper.h:
```diff
@@ -75,6 +83,7 @@ class CDBBatch
         leveldb::Slice slValue(ssValue.data(), ssValue.size());
 
         batch.Put(slKey, slValue);
+        size_estimate += 3 + slKey.size() + slValue.size();
```
Done

sipa commented:
Addressed some of @TheBlueMatt's comments.

TheBlueMatt commented on src/dbwrapper.h:
```diff
@@ -75,6 +83,14 @@ class CDBBatch
+        // - byte[]: key
+        // - varint: value length
+        // - byte[]: value
+        // The formula below assumes the key and value are both less than 16k.
```
Wait, is this wrong? Looks like you're assuming both varints to be 1 byte, which the comment suggests is only up to 127B, not 16KB?

sipa commented on src/validation.cpp:
```diff
@@ -3681,6 +3702,51 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    if (mapBlockIndex.count(hashBest)) {
+        auto pindexBest = mapBlockIndex[hashBest];
+        if (pindexUpto->GetAncestor(pindexBest->nHeight) != pindexBest) {
+            return error("ReplayBlocks(): chainstate tip does not derive from final boundary");
```
That's a great point. I hadn't considered that in the case of a reorg the set of partially written changes may include things from multiple branches. It seems solvable by allowing the 'upto' blocks to be a list of tip hashes, and then at reply time choose which ones to undo and which ones to replay. I think that's a problem for later, but it makes sense to have a comment about it.

sipa replied:
Good catch. ReplayBlocks should learn to deal with a reorg.

sipa commented on src/txdb.cpp:
```diff
@@ -44,10 +47,21 @@ uint256 CCoinsViewDB::GetBestBlock() const {
     size_t changed = 0;
+    size_t batch_size = (size_t)GetArg("-dbbatchsize", nDefaultDbBatchSize) << 20;
+    if (!hashBlock.IsNull()) {
+        batch.Write(DB_BEST_BLOCK_UPTO, hashBlock);
```
That's a great point. I hadn't considered that in the case of a reorg the set of partially written changes may include things from multiple branches. It seems solvable by allowing the 'upto' blocks to be a list of tip hashes, and then at start time choose which ones to undo and which ones to replay. I think that's a problem for later, but it makes sense to have a comment about it.

sipa committed:
[MOVEONLY] Move LastCommonAncestor to chain

sipa commented:
Updated to deal with reorganizations. The disk format and recovery code can now also deal with multiple partially written branches. That functionality is not needed yet, but means we can switch to different partial flushing strategies later without breaking compatibility with older versions.

TheBlueMatt committed:
Dont create pcoinsTip until after ReplayBlocks.

This requires that we not access pcoinsTip in InitBlockIndex's
FlushStateToDisk (so we just skip it until later in AppInitMain)
and the LoadChainTip in LoadBlockIndex (which there is already one
later in AppinitMain, after ReplayBlocks, so skipping it there is
fine).

Includes some simplifications by Suhas Daftuar and Pieter Wuille.

TheBlueMatt commented on src/init.cpp:
```diff
@@ -1476,6 +1479,13 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
+                    strLoadError = _("Unable to replay blocks. You will need to rebuild the database using -reindex-chainstate.");
+                    break;
+                }
+                pcoinsTip->SetBestBlock(pcoinsdbview->GetBestBlock()); // TODO: only initialize pcoinsTip after ReplayBlocks
```
I think you can do this for (almost) free now. See https://github.com/TheBlueMatt/bitcoin/commit/747b766a02c32a47a5f0018ed352991b55aaab19, though if you dont want to take it here I'll just PR it afterwards.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3690,6 +3693,124 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    if (blockUndo.vtxundo.size() + 1 != block.vtx.size())
+        return error("RollbackBlock(): block and undo data inconsistent");
+
+    for (size_t i = 0; i < block.vtx.size(); ++i) {
```
Doesn't this need to be in reverse order (like in DisconnectBlock, maybe you should just go ahead and add an option to DisconnectBlock to ignore errors in a pervious commit to make it easier to review?)?

sipa replied:
Good catch, fixed.

I'll try to do the merging in a extra commit.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1505,18 +1505,15 @@ bool ApplyTxInUndo(const CTxInUndo& undo, CCoinsViewCache& view, const COutPoint
         // undo data contains height: this is the last output of the prevout tx being spent
-        if (!coins->IsPruned())
-            fClean = fClean && error("%s: undo data overwriting existing transaction", __func__);
+        if (!coins->IsPruned()) fClean = false;
```
Would be nice to not lose the error messages by adding a flag for printing errors. I think you need the flag either way for the next line, as I dont think you can run the Clear() if we're re-undoing a tx.

TheBlueMatt replied:
I just find the errors useful to keep around, my real concern is the Clear() one line down, which I believe is an actual bug for the new usage in RollbackBlock.

sipa replied:
I'm not convinced the Clear() is wrong - it does mean we're passing over a state where the outputs for that TX were all full spent. However, this is hard to reason about, and you may well be right. Furthermore, it seems that Clear() has no purpose. In the 'clean' case, the output is already pruned, to the Clear() is a no-op. In the other case it doesn't matter. I'm removing it.

TheBlueMatt replied:
Hmm, you may be right. Indeed, however, hard to reason about.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -3690,6 +3693,124 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    std::set<const CBlockIndex*, CBlockIndexWorkComparator> vpindexRollback;
+    for (size_t i = 1; i < pindexHeads.size(); ++i) {
+        const CBlockIndex *pindexHead = pindexHeads[i];
+        while (fGenesis ? pindexHead != nullptr : pindexHead != pindexFork) {
```
I'm not convinced this is right. What if you connect both A and B, in the simple case? Now you'll disconnect B before you re-connect A and then re-connect B? Is that neccessary?

sipa replied:
Note that i starts at 1 in the loop, skipping the branch that leads to the new tip. I've added a comment to clarify.

TheBlueMatt replied:
Indeed, but that only means you wont do the disconnect-then-reconnect thing for one block (which I suppose may be fine for this PR), but you will do it if you have two back-to-back blocks in the list (or am I confused?).

sipa replied:
No. Let's say you have a chain A->B->C that was being flushed (meaning the old tip was A, you crashed in the middle of writing the changes for B and C, with C the intended new tip). In this case, at recovery time, GetBlockHeads() will return `[C,A]`. `pindexFork` will be A. The loop above will only process A, but because A is already the fork point, nothing is added to the disconnect set.

TheBlueMatt replied:
Yes, OK, it wasnt clear to me what GetBlockHeads() should be returning there, Flush seemed to indicate something different.

sipa reviewed (commented):
Thanks for the detailed review, @TheBlueMatt.

NicolasDorier commented on src/txdb.cpp:
```diff
@@ -44,10 +45,33 @@ uint256 CCoinsViewDB::GetBestBlock() const {
+        std::vector<uint256> heads = GetHeadBlocks();
+        // Construct a set with all existing heads, excluding the new tip.
+        std::set<uint256> setHeads(heads.begin(), heads.end());
+        if (setHeads.empty() || !tip.IsNull()) setHeads.insert(tip);
```
can `tip` be null ? I would have expected it is at least genesis.

sipa replied:
Tip can be null when it's the first write ever to the database.

NicolasDorier replied:
I am surprised, I always thought the coinbase of the genesis block was not added in the database. This would mean that having it to `null` to say `before processing the genesis`, is the same than enforcing the best block to be at least the genesis block.

NicolasDorier commented on src/validation.cpp:
```diff
@@ -3686,6 +3691,124 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+
+    // Find last common ancestor of all heads.
+    for (const uint256& hash : hashHeads) {
+        if (hash.IsNull()) {
```
I am confused about all this `fGenesis` variable. I guess this is related to my other comment on https://github.com/bitcoin/bitcoin/pull/10148/commits/02dfa4a54da44e40b1d29687cc68885f3a306ef3#r111875593

I would not expect, pindexFork to be null, as the work fork point should be the genesis block.

sipa replied:
I've added a length comment to clarify.

laanwj commented:
> contrib/devtools/check-doc.py is unhappy that you added new arguments without asking for permission from the argument gods.

Not sure if you were being sarcastic here, but to be clear: that script checks whether options are documented (either as debug option or as normal option), not whether you have a signed permission on a stone tablet from $PANTHEON.

Going to test this on a node.

sipa committed:
Non-atomic flushing using the blockchain as replay journal

sipa committed:
Adapt memory usage estimation for flushing

sipa commented:
Rebased, and removed the multi-reorg support in the recovery code.

sdaftuar commented:
@sipa Thanks for the simplification.  I've read through the logic and I believe it is correct; will think about ways to carefully test.

There was some discussion on IRC but I wanted to summarize my current thoughts: this is still some complication we're adding to consensus, and I expect that the per-txout caching from #10195 will change the behavior and performance of the cache substantially, so I'd prefer to review that PR independently of this one to better consider the design decisions of this PR.

sipa commented:
@sdaftuar My plan is adding a unit test that uses wrapper around CCoinsViewDB that drops 50% of the dirty entries before writing, and then issues ReplayBlock to recover from it.

NicolasDorier commented on src/validation.cpp:
```diff
@@ -3691,6 +3691,86 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+            LogPrintf("Rolling back %s (%i)\n", pindexOld->GetBlockHash().ToString(), pindexOld->nHeight);
+            bool fClean = true; // Ignoring resulting state of fClean.
+            CValidationState state;
+            if (!DisconnectBlock(block, state, pindexOld, cache, &fClean)) return false;
```
maybe return an `error()` here ?

NicolasDorier commented on src/validation.cpp:
```diff
@@ -3691,6 +3691,86 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+            return error("ReplayBlocks(): reorganization from unknown block requested");
+        }
+        pindexOld = mapBlockIndex[hashHeads[1]];
+        pindexFork = LastCommonAncestor(pindexOld, pindexNew);
```
I would add a `assert(pindexFork);`

NicolasDorier commented:
CodeReview ACK. Would be more at ease with some functional tests.

An easy test is trying to sync a node for 1000 blocks, with dbbatchsize to 1 bytes (it is in MB right now though) and dbcrashratio to like 10, and see if it eventually manage to do it.

NicolasDorier commented:
Maybe an easier way to achieve the same result would be to move the estimated size of a Batch to the `CCoinsViewCache` and add a method `CCoinsViewCache::FlushIfNeeded()` that the caller will call when he does not really need to flush.

morcos commented on src/txdb.cpp:
```diff
@@ -59,12 +80,22 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
         mapCoins.erase(itOld);
+        if (batch.SizeEstimate() > batch_size) {
+            LogPrint(BCLog::COINDB, "Writing partial batch of %.2f MiB\n", batch.SizeEstimate() * (1.0 / 1048576.0));
+            db.WriteBatch(batch);
```
Should we be looking for failure here and returning false if any of the `WriteBatch` calls fail?

sipa replied:
WriteBatch throws a dbwrapper_error when writing fails.

TheBlueMatt commented on src/txdb.cpp:
```diff
@@ -59,12 +82,30 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
+            LogPrint(BCLog::COINDB, "Writing partial batch of %.2f MiB\n", batch.SizeEstimate() * (1.0 / 1048576.0));
+            db.WriteBatch(batch);
+            batch.Clear();
+            int crash_simulate = GetArg("-dbcrashratio", 0);
```
Can we move GetArg lookup out of the hot loop?

sipa replied:
Done.

TheBlueMatt reviewed (commented):
utACK 29548bcdf9130bb40f164c20692fcbfcfd42c109 minus one GetArg and needing squashing.

morcos commented:
Concept ACK.  I've done a preliminary review and it looks pretty good, but I'd like to spend some more time thinking about it.

sipa commented:
Rebased on #10195.

gmaxwell commented:
@sipa Needs rebase!

sipa commented:
Rebased after the merge of #10195.

gmaxwell commented:
@TheBlueMatt you didn't want to finish a review of this before until after the per-txo changes happened. They've happened.   Time to turn over your soul^wreview!

sdaftuar commented:
Needs rebase.  Also is it time now to remove WIP tag?

sipa commented:
Rebased.

ryanofsky commented on src/validation.cpp:
```diff
@@ -3535,6 +3541,88 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    int nForkHeight = pindexFork ? pindexFork->nHeight : 0;
+    for (int nHeight = nForkHeight + 1; nHeight <= pindexNew->nHeight; ++nHeight) {
+        const CBlockIndex* pindex = pindexNew->GetAncestor(nHeight);
+        LogPrintf("Rolling forward %s (%i)\n", pindex->nHeight, nHeight);
```
%s could be %i.

Also, is there a case where `pindex->nHeight` would not be the same as `nHeight`?

sipa replied:
Fixed.

sipa commented on src/validation.cpp:
```diff
@@ -3535,6 +3541,88 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+            if (res == DISCONNECT_FAILED) {
+                return error("RollbackBlock(): DisconnectBlock failed at %d, hash=%s", pindexOld->nHeight, pindexOld->GetBlockHash().ToString());
+            }
+            // We can safely ignore unclean here.
```
Added a comment.

sipa commented on src/txdb.cpp:
```diff
@@ -68,10 +69,30 @@ uint256 CCoinsViewDB::GetBestBlock() const {
+        // In the first batch, mark the database as being in the middle of a
+        // transition from old_tip to hashBlock.
+        batch.Erase(DB_BEST_BLOCK);
+        batch.Write(DB_HEAD_BLOCKS, std::vector<uint256>{hashBlock, old_tip});
```
It's for future compatibility with a design where entries are continuously written to disk in the background, which may lead to multiple partially-applied reorgs to co-exist on disk.

sdaftuar reviewed (commented):
Re-reviewed; it looks pretty good though I have a couple questions...   Will work on testing.

sipa commented on src/txdb.cpp:
```diff
@@ -68,10 +69,32 @@ uint256 CCoinsViewDB::GetBestBlock() const {
 
+std::vector<uint256> CCoinsViewDB::GetHeadBlocks() const {
+    std::vector<uint256> vhashHeadBlocks;
+    if (!db.Read(DB_HEAD_BLOCKS, vhashHeadBlocks))
```
Fixed.

sipa commented on src/txdb.cpp:
```diff
@@ -68,10 +69,32 @@ uint256 CCoinsViewDB::GetBestBlock() const {
+        if (old_tip.IsNull()) {
+            // We may be in the middle of replaying.
+            std::vector<uint256> old_heads = GetHeadBlocks();
+            if (old_heads.size() >= 2) old_tip = old_heads[1];
```
Done.

sipa commented on src/txdb.cpp:
```diff
@@ -84,10 +107,19 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock) {
+        }
     }
-    if (!hashBlock.IsNull())
+    if (!hashBlock.IsNull()) {
```
I've removed the need for flushing with hash=0.

sipa commented on src/validation.cpp:
```diff
@@ -3535,6 +3541,91 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+    return true;
+}
+
+bool ReplayBlocks(const CChainParams& params, CCoinsView* view)
```
I prefer to keep types as generic as they can be; perhaps some caching layer is later added.

sipa commented on src/init.cpp:
```diff
@@ -1494,9 +1493,13 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                 }
-                pcoinsTip->SetBestBlock(pcoinsdbview->GetBestBlock()); // TODO: only initialize pcoinsTip after ReplayBlocks
+                pcoinsTip = new CCoinsViewCache(pcoinscatcher);
+                pcoinsTip->SetBestBlock(pcoinsdbview->GetBestBlock());
```
Indeed, removed.

sipa commented on src/init.cpp:
```diff
@@ -1494,9 +1493,13 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                 LoadChainTip(chainparams);
 
+                // Force a chainstate write so that when we VerifyDB in a moment, it doesn't check stale data
+                FlushStateToDisk();
```
Indeed, removed (and this also allows removing the hash=0 flushing logic).

sdaftuar committed:
[qa] Test non-atomic chainstate writes

Adds new functional test, dbcrash.py, which uses -dbcrashratio to exercise the
logic for recovering from a crash during chainstate flush.

dbcrash.py is added to the extended tests, as it may take ~10 minutes to run

Use _Exit() instead of exit() for crash simulation

This eliminates stderr output such as:
    terminate called without an active exception
or
    Assertion failed: (!pthread_mutex_destroy(&m)), function ~recursive_mutex, file /usr/local/include/boost/thread/pthread/recursive_mutex.hpp, line 104.

Eliminating the stderr output on crash simulation allows testing with
test_runner.py, which reports a test as failed if stderr is produced.

sdaftuar reviewed (commented):
It occurs to me that we aren't bounding memory usage at all during recovery.  This seems like it could cause problems for users who are running with a big -dbcache and then crash.

Perhaps we should suggest a -reindex if there are a lot of blocks to disconnect/connect, or abort and suggest a -reindex if the cache in ReplayBlocks exceeds the  dbcache+mempool size?

sdaftuar commented:
Also, I was able to write a functional test that exercises this logic (and which caught the `undo.fCoinbase` vs `!fClean` issue in `ApplyTxInUndo()`):
https://github.com/sdaftuar/bitcoin/commits/test-10148

sipa commented on src/txdb.h:
```diff
@@ -74,6 +76,7 @@ class CCoinsViewDB : public CCoinsView
     bool GetCoin(const COutPoint &outpoint, Coin &coin) const override;
     bool HaveCoin(const COutPoint &outpoint) const override;
     uint256 GetBestBlock() const override;
+    std::vector<uint256> GetHeadBlocks() const;
```
Fixed.

sipa commented on src/validation.cpp:
```diff
@@ -1328,17 +1328,15 @@ int ApplyTxInUndo(Coin&& undo, CCoinsViewCache& view, const COutPoint& out)
         }
     }
-    view.AddCoin(out, std::move(undo), undo.fCoinBase);
+    view.AddCoin(out, std::move(undo), !fClean);
```
Added some comments.

sipa commented:
* Made the added argument to `AddCoins` `check` instead of `overwrite`. Its semantics are now to issue a `HaveCoin` to determine whether an overwrite happens, rather than just assume there is one. This has the advantage that (some) of the coins added in `RollForwardBlock` correctly get a `FRESH` marker, possibly reducing memory usage if the rollforward is large.
* Added @sdaftuar 's dbcrash test (which still runs succesfully).

@sdaftuar As for memory usage during recovery, I think the correct approach is to make the recovery itself non-atomic later.

sdaftuar commented:
Looks good, ACK.  However, if we include the dbcrash test, then I think we also need something like https://github.com/sdaftuar/bitcoin/commit/96a7ef4e990cb33f4e5cc248b630d17b2384d33b in order for `test/functional/test_runner.py --extended` to not fail?

sdaftuar commented:
ACK 03604ea5f3de4dd3301533ab353b2490105ebdb0

laanwj commented:
Small merge conflict in `check-doc.py`

gmaxwell commented:
@sdaftuar Thats a great test.

@sipa needs conflicts fixed

sipa commented:
Rebased, and squashed the last two commits.

sipa commented:
Travis where art thou

sdaftuar commented:
re-ACK 176c021d085f5a45bc9e038e760942aa648dd797

laanwj commented:
Tested ACK 176c021

TheBlueMatt commented:
utACK-sans-tests once the fixes for init-order bugs here goes through in #10758.

morcos commented:
posthumous utACK-sans-tests and modulo some of the same bugs @TheBlueMatt found and fixed in #10758.  I'll review that now.
